packages/aes-kw/aes-kw.ts

```diff
@@ -0,0 +1,138 @@
+     * constructor
+     * @param keyData: The key data to wrap with the key encryption key
+     */
+    wrapKey(keyData: Uint8Array): Uint8Array {
+        // Floor divide the length of the key data to determine 
+        // how many 64 bit data blocks it contains
+        const N = (keyData.length - (keyData.length % 8)) / 8;
+
+        // The keyData must be at minimum 128 bit (16 bytes) in length
+        if (N <= 1) {
+            throw new Error("aeskw: key data insufficient length must be a minimum of 128 bits or 16 bytes");
+        }
+
+        // Set A to the initial value
+        let A = this._iv;
+        // Initialize the length of the wrapped key, size always equals N+1
+        const wrappedKey = new Uint8Array(8*(N+1));
+        // Set the plain text into the wrapped key array offset by one
+        // 64 bit data block
+        wrappedKey.set(keyData, 8);
+
+        for (let j = 0; j < 6; j++) {
+            for (let i = 1; i <= N; i++) {
+                this._inputBuffer.set(A);
+                this._inputBuffer.set(wrappedKey.subarray(i*8,(i+1)*8),8);
+                this._cipher.encryptBlock(this._inputBuffer, this._outputBuffer);
+                A = writeUint64BE(i + j*N);
+                this.xor(A, this._outputBuffer.subarray(0,8));
+                wrappedKey.set(this._outputBuffer.subarray(8,16), i*8);
+            }
+        }
+
+        wrappedKey.set(A);
+        return wrappedKey;
+    }
+
+    /**
+     * Un-wraps a wrapped key using the key encryption key supplied in the
+     * constructor
+     * @param wrappedKey: The wrapped key to un-wrap with the key encryption key
+     */
+    unwrapKey(wrappedKey: Uint8Array): Uint8Array {
+        // Floor divide the length of the unwrapped key to determine 
+        // how many 64 bit data blocks it contains, N represents the number
+        // of 64 bit data blocks of the plain text which is always one less
+        // than the cipher text
+        const N = ((wrappedKey.length - (wrappedKey.length % 8)) / 8) - 1;
+        
+        // The keyData must be at minimum 128 bit (16 bytes) in length
+        if (N <= 1) {
+            throw new Error("aeskw: key data insufficient length must be a minimum of 128 bits or 16 bytes");
+        }
+
+        // Set A to the first 64 bit data block of the wrapped key
+        let A = wrappedKey.subarray(0,8);
+        // Initialize the length of the key data, size always equals N
+        const keyData = new Uint8Array(8*(N));
+
+        for (let j = 5; j >= 0; j--) {
+            for (let i = N; i >= 1; i--) {
+                this.xor(A, writeUint64BE(i + j*N));
+                this._inputBuffer.set(A);
+                this._inputBuffer.set(wrappedKey.subarray((i)*8,(i+1)*8),8);
+                this._cipher.decryptBlock(this._inputBuffer, this._outputBuffer);
+                A.set(this._outputBuffer.subarray(0,8));
+                wrappedKey.set(this._outputBuffer.subarray(8,16), i*8);
```

Comment:
I don't like that it changes the argument. Could you please copy `wrappedKey` in the beginning of the method (after error handling), something like:

```
const tempKey = new Uint8Array(wrappedKey)
```

and then use it here and in `let A = ...`.

Or maybe call it `keyData` and then slice off the first 8 bytes at the end.

Reply:
Yeap originally I had this but have re-added